### PR TITLE
Polyspear desync and better battle summary

### PR DESCRIPTION
**Pull request by PieroganoReggiano**
ready, but please don't merge -- i want to resolve conflicts and push merge commit myself

### CODE REVIEW

File: Scripts/Battle/battle_manager.gd [hunk 1]
```diff
@@ -127,7 +136,25 @@ func start_battle(new_armies : Array[Army], battle_map : DataBattleMap, \
...
 	battle_ui.load_armies(_battle_grid.armies_in_battle_state)
 
-	on_turn_started(_battle_grid.get_current_player())
+	if not _batch_mode:
+		on_turn_started(_battle_grid.get_current_player())
+
+
+func force_battle_state(armies : Array[Army], \
+		map_data : DataBattleMap, \
+		battle_state : SerializableBattleState, x_offset : float):
```
Comment (NamespaceV) [inline Scripts/Battle/battle_manager.gd:145]: i would not add x_offset as a parameter
x_offset is hacky way to make battles not collide with world, its not a real parameter

also shouldn't SerializableBattleState contain map and armies...
seams weerd to have separate parameters

File: Scripts/Battle/battle_manager.gd [hunk 2]
```diff
@@ -152,12 +179,20 @@ func replay_move_delay() -> void:
...
+	return result
+
+
 #endregion
 
 
 #region Input Functions
 
-## Currently only used for AI
+## Currently only used for AI -- sure?
```
Comment (NamespaceV) [inline Scripts/Battle/battle_manager.gd:195]: AFAIK yes, should be renamed

File: Scripts/Battle/battle_manager.gd [hunk 3]
```diff
@@ -454,23 +506,23 @@ class AnimInQueue:
 		result.debug_name = "turn_"+unit_form_.unit.template.unit_name
 		result._unit_form = unit_form_
 		unit_form_.anim_end.connect(result.on_anim_end)
-		result._animate = func () : unit_form_.start_turn_anim()
+		result._animate = func () : if (unit_form_): unit_form_.start_turn_anim()
```
Comment (NamespaceV) [inline Scripts/Battle/battle_manager.gd:509]: why is this needed?
if you try animating null unit_form_ it will break on not getting to anim end ever
null check doesn't help much

File: Scripts/General/input_manager.gd [hunk 4]
```diff
@@ -29,8 +29,11 @@ func get_battle_maps_list() -> Array[String]:
 	return FileSystemHelpers.list_files_in_folder(CFG.BATTLE_MAPS_PATH)
 
 
-## starts game based on game_setup_info
-func start_game():
+func start_or_set_game_in_state(world_state : SerializableWorldState, \
```
Comment (NamespaceV) [inline Scripts/General/input_manager.gd:32]: i'd make separate functions,
duplication is low, just separate clear_players to a separate function
there is no point in having such huge if else and force_state check

File: Scripts/Multiplayer/NET_manager.gd [hunk 5]
```diff
@@ -11,6 +11,11 @@ signal chat_message_arrived(content : String)
 signal chat_log_cleared
 
 
+func _process(_delta) -> void:
+	if Input.is_action_just_pressed("FORCE_DESYNC"):
```
Comment (NamespaceV) [inline Scripts/Multiplayer/NET_manager.gd:15]: I'd keep input in input manager
having `Input.is_action_just_pressed` scattered around code base is messy

File: Scripts/World/Places/place.gd [hunk 6]
```diff
@@ -62,3 +63,24 @@ func get_map_description() -> String:
...
+	# var script : String = get_script().resource_path.get_file()
+	# dict["script"] = script
+	return dict
+
+
+## should be overridden by each place
+## but also this is a temporary part of greater refactor which is needed for
+## Place class
+func to_specific_serializable(dict : Dictionary) -> void:
+	dict["not implemented"] = true
```
Comment (NamespaceV) [inline Scripts/World/Places/place.gd:86]: `assert(false, "to_specific_serializable not implemented")` if doesn't crash

File: Scripts/World/world_grid_node2d.gd [hunk 7]
```diff
@@ -20,14 +21,16 @@ func load_map(world_map : DataWorldMap) -> void:
 		for y in range(grid_height):
 			var coord := Vector2i(x, y)
 			var data : DataTile = world_map.grid_data[x][y]
-			var place : Place = Place.create_place(data, coord)
+			var place : Place = null
+			if not state_load_mode or true:
```
Comment (NamespaceV) [inline Scripts/World/world_grid_node2d.gd:25]: convoluted `if true:` ???

File: Scripts/World/world_manager.gd [hunk 8]
```diff
@@ -24,10 +24,19 @@ var current_player : Player
...
 
+var _batch_mode : bool = false
+
 #endregion
 
 signal world_move_done
 
+
+func _process(_delta):
+	if Input.is_action_just_pressed("PRINT_WORLD_STATE"):
```
Comment (NamespaceV) [inline Scripts/World/world_manager.gd:35]: as earlier, i'd keep all Input.is_action_just_pressed in input manager
so there is a single place to check for key bindings tweaks
not whole code base

Comment (NamespaceV) [reply Scripts/World/world_manager.gd:35]: with IM.quit_game() doubly concerning

File: Scripts/World/world_manager.gd [hunk 9]
```diff
@@ -75,6 +84,10 @@ func get_player_index(player : Player) -> int:
 
 
 func get_player_by_index(index : int) -> Player:
+	if index < 0:
+		return null
```
Comment (NamespaceV) [inline Scripts/World/world_manager.gd:88]: is this needed in a legitimate case or just hides errors to be fixed?
maybe a `push_warning`

**Review by NamespaceV — APPROVED** [review]

File: Scripts/Battle/battle_manager.gd (hunk 1 above)
Comment (NamespaceV) [reply Scripts/Battle/battle_manager.gd:145]: ok this is not a force_battle_state
it is alternative start_battle with state to deserialize...

Comment (PieroganoReggiano) [reply Scripts/Battle/battle_manager.gd:145]: The parameter `armies` is separate because battle state is optional and has these values in other format. `map_data` is in same format, good point. I didn't have better idea yet for it, so I left it this way. Probably I will fully merge start_battle and force_battle_state, but I'm not sure yet. Also, these parameters like `world_armies` etc. can be rearranged.

File: Scripts/Battle/battle_manager.gd (hunk 3 above)
Comment (PieroganoReggiano) [reply Scripts/Battle/battle_manager.gd:509]: It is needed when we quit battle when any animation is playing -- the clean up i added is not sufficient and we need to do additional null check

File: Scripts/General/input_manager.gd (hunk 4 above)
Comment (PieroganoReggiano) [reply Scripts/General/input_manager.gd:32]: some small rearrange

File: Scripts/Multiplayer/NET_manager.gd (hunk 5 above)
Comment (PieroganoReggiano) [reply Scripts/Multiplayer/NET_manager.gd:15]: xd, it is temporary debug which will not go to master

File: Scripts/World/Places/place.gd (hunk 6 above)
Comment (PieroganoReggiano) [reply Scripts/World/Places/place.gd:86]: idk, i will leave it for now as it is -- `Place` needs rework

File: Scripts/Multiplayer/NET_manager.gd (hunk 5 above)
Comment (NamespaceV) [reply Scripts/Multiplayer/NET_manager.gd:15]: i'd keep this one it can be useful for testing, maybe as a cheat in the chat

File: Scripts/Battle/battle_manager.gd (hunk 2 above)
Comment (NamespaceV) [reply Scripts/Battle/battle_manager.gd:195]: lol AI and UI a few lines below
I'll move this comment when refactoring BM

File: Scripts/World/world_manager.gd (hunk 9 above)
Comment (PieroganoReggiano) [reply Scripts/World/world_manager.gd:88]: neutrals are encoded as -1

File: Scripts/World/world_grid_node2d.gd (hunk 7 above)
Comment (PieroganoReggiano) [reply Scripts/World/world_grid_node2d.gd:25]: looks like an accident

**Comment by PieroganoReggiano** [comment]
ready -- I simplified many of code chunks I worked -- can be merged

File: Scripts/World/world_manager.gd (hunk 9 above)
Comment (NamespaceV) [reply Scripts/World/world_manager.gd:82]: i suggest we add 
`const NEUTRAL_PLAYER_INDEX = -1`
and check for it here
